x/concentrated-liquidity/swaps_test.go

```diff
@@ -2528,23 +2580,115 @@ func (s *KeeperTestSuite) TestInverseRelationshipSwapInAmtGivenOut() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestUpdatePoolForSwap() {
+	var (
+		oneHundredETH         = sdk.NewCoin(ETH, sdk.NewInt(100_000_000))
+		oneHundredUSDC        = sdk.NewCoin(USDC, sdk.NewInt(100_000_000))
+		defaultInitialBalance = sdk.NewCoins(oneHundredETH, oneHundredUSDC)
+	)
+
+	tests := map[string]struct {
+		senderInitialBalance sdk.Coins
+		poolInitialBalance   sdk.Coins
+		tokenIn              sdk.Coin
+		tokenOut             sdk.Coin
+		newCurrentTick       sdk.Int
+		newLiquidity         sdk.Dec
+		newSqrtPrice         sdk.Dec
+		expectError          error
+	}{
+		"success case": {
+			senderInitialBalance: defaultInitialBalance,
+			poolInitialBalance:   defaultInitialBalance,
+			tokenIn:              oneHundredETH,
+			tokenOut:             oneHundredUSDC,
+			newCurrentTick:       sdk.NewInt(2),
+			newLiquidity:         sdk.NewDec(2),
+			newSqrtPrice:         sdk.NewDec(2),
+		},
+		"sender does not have enough balance": {
+			senderInitialBalance: defaultInitialBalance,
+			poolInitialBalance:   defaultInitialBalance,
+			tokenIn:              oneHundredETH.Add(oneHundredETH),
+			tokenOut:             oneHundredUSDC,
+			newCurrentTick:       sdk.NewInt(2),
+			newLiquidity:         sdk.NewDec(2),
+			newSqrtPrice:         sdk.NewDec(2),
+			expectError:          types.InsufficientUserBalanceError{},
+		},
+		"pool does not have enough balance": {
+			senderInitialBalance: defaultInitialBalance,
+			poolInitialBalance:   defaultInitialBalance,
+			tokenIn:              oneHundredETH,
+			tokenOut:             oneHundredUSDC.Add(oneHundredUSDC),
+			newCurrentTick:       sdk.NewInt(2),
+			newLiquidity:         sdk.NewDec(2),
+			newSqrtPrice:         sdk.NewDec(2),
+			expectError:          types.InsufficientPoolBalanceError{},
+		},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		suite.Run(name, func() {
+			suite.SetupTest()
+			concentratedLiquidityKeeper := suite.App.ConcentratedLiquidityKeeper
+
+			// Create pool with initial balance
+			pool := suite.PrepareConcentratedPool()
+			suite.FundAcc(pool.GetAddress(), tc.poolInitialBalance)
+
+			// Create account with empty balance and fund with initial balance
+			sender := apptesting.CreateRandomAccounts(1)[0]
+			suite.FundAcc(sender, tc.senderInitialBalance)
+
+			// Default pool values are initialized to one.
+			pool.ApplySwap(sdk.OneDec(), sdk.OneInt(), sdk.OneDec())
+
+			err := concentratedLiquidityKeeper.UpdatePoolForSwap(suite.Ctx, pool, sender, tc.tokenIn, tc.tokenOut, tc.newCurrentTick, tc.newLiquidity, tc.newSqrtPrice)
+
+			if tc.expectError != nil {
+				suite.Require().Error(err)
+				suite.Require().ErrorAs(err, &tc.expectError)
+				return
```

Comment:
Shouldn't we be checking the pool does not change in the error case?

Reply:
[b76748b](https://github.com/osmosis-labs/osmosis/pull/4533/commits/b76748be5ad31e7079425cc991124258263f4114)